Add details of our SLSA level 2 compliance. (#202)
adoptium/blog.adoptium.net: content/blog/eclipse-temurin-slsa2/index.md

  releases to guarantee their authenticity.

- **Provenance - service generated**. The GPG signatures for the binary
  are generated as an integral part of the build process for the binaries
  and will be done in the same way for the signing of the SBOMs - the current
  SBOM signing was done retrospectively in our Jenkins instance. The Jenkins
  instance hosts the private keys from the Eclipse Foundation which are used
  to perform the signature generation.

# Next steps for the project

Are we finished? Absolutely not! We are continuing to work towards
achieving the higher levels of SLSA and have already achieved many of the
requirements of higher levels. But for now, we are proud to be able to claim
full compliance with SLSA level 2.

We are currently tracking forward progress to meeting the higher level
requirements at
[adoptium#160](https://github.com/adoptium/adoptium/issues/160). We already
meet some of the requirements of higher levels - in some cases there are
differences in the criteria which we meet on different platforms. For
example, we have worked to [make our build process reproducible](https://blog.adoptium.net/2022/06/adoptium-reproducible-builds/)

> and on Linux, Windows and MacOS for JDK19+ we are able to produce binary

Reviewer: Can it now be claimed as 17+ ?

Author: I wasn't clear that everythign has been backported on all platforms yet so I'll defer to @andrew-m-leonard on that.  It can always be updated later on as things move forward ;-)

Reviewer 2: jdk17 yes, all backports from 19 to 17 that are needed are done. but of course not jdk18, so jdk17+ is not quite accurate, it should be:
**jdk17, and jdk19+**

Author: @andrew-m-leonard Does that mean that the latest 17.0.5+8 release was reproducible on all three operating systems?

Reviewer 2: i'm going to draw up a "table" which should make it clearer, also got plans on possible nightly build-repo...

Author: reworded to `on Linux, Windows and MacOS for JDK17 (LTS) and JDK19+ we have created
and shipped binary reproducible builds`

---
title: SLSA level 2 compliance for Eclipse Temurin

> date: "2022-11-11T12:00:00+00:00"

Author: Updated to today's date for now :-)

author: sxa
description: Eclipse Temurin by Adoptium is compliant with level 2 of the SLSA secure development framework.
tags:
  - Temurin
  - Security
---

# Introduction to SLSA

> [SLSA](https://slsa.dev) is a framework with individual levels to indicate
> increasing levels of security which software products can work towards. 

Reviewer 3: IMHO the SLSA website is a bit confused about sometimes addressing the supply chain vs. defining a level of general software development/consumption security. It certainly complements our wider security efforts. It may be better to paraphrase the purpose as described on their website to avoid a misunderstanding.

```suggestion
[SLSA](https://slsa.dev) is a framework with individual levels that software producers can work towards to make their software more secure, and consumers can make decisions based on the software package’s security posture. 
```

---
title: SLSA level 2 compliance for Eclipse Temurin
date: "2022-11-11T12:00:00+00:00"
author: sxa
description: Eclipse Temurin by Adoptium is compliant with level 2 of the SLSA secure development framework.
tags:
  - Temurin
  - Security
---

# Introduction to SLSA

> [SLSA](https://slsa.dev) is a framework with inndividual levels to indicate

Reviewer: ```suggestion
[SLSA](https://slsa.dev) is a framework with individual levels to indicate
```

  jenkins pipelines from the [ci-jenkins-pipelines](https://github.com/adoptium/ci-jenkins-pipelines) repository and the
  underlying scripts run on the build machines from the
  [temurin-build](https://github.com/adoptium/temurin-build)
  repository - starting at
  [make-adopt-build-farm.sh](https://github.com/adoptium/temurin-build/blob/master/build-farm/make-adopt-build-farm.sh)

- **Provenance - available**: We create an SBOM in OWASP CycloneDX format
  along with our releases which contains all of the information about how
  the builds are produced, which should allow someone to rebuild if required
  for any reason.  This includes the full set of parameters which we use to
  the makejdk_any_platform.sh script (invoked from
  make-adopt-build-farm.sh), the source repository tags (`scmRef`) which was
  used by the build process to produce the builds, the output from the
  openjdk `configure` invocation and various other pieces of information. 
  We are continually evolving the specific details which we include in the
  SBOM.  If you want to join in the discussion on the content, you can find
  the conversation in

>   [temurin-build#3013](https://github.com/adoptium/temurin-build/issues/3013)

Reviewer: Or join the conversation in our Slack workspace in the secure-dev channel

Author: Added a line to the same effect.

## SLSA Level 2

Level 2 adds in additional requirements to provide some tamper resistance of
the build process, mostly in the areas of having all of our code version
controlled which we do by storing anything related to producing our buids in
git in our case and also by using a build service to produce and distribute
those builds - in our case this is our jenkins CI server. In our release
cycles since July 2022 we have also provided PGP signature verification
information along with the SHA checksums that we previously had, which was
the final part of achieving SLSA level 2 compliance. The requirements here
are:

- **Source - Version controlled**. All of our source code is stored in GitHub in
  the jdkXX repositories such as
  [jdk17u](https://github.com/adoptium/jdk17u/), which we mirror from the
  [openjdk project](https://github.com/openjdk). These
  are version controlled and have tags for each release which we build from.
  Contributors agree to the rules of the
  [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

> - **Build - Build service**. All of the build steps are run using our jenkins

Author: Done - and made the change elsewhere in the doc too.

  build service - currently at https://ci.adoptopenjdk.net which is used to
  perform the builds, generate the SBOMs, and build the installers where
  applicable. The output from the builds are then posted into GitHub release
  repositories named as temurinXX-binaries (e.g.
  [temurin17-binaries](https://github.com/adoptium/temurin17-binaries) and
  also exposed via our API and web site.

- **Provenance - Authenticated**. We provide GPG checksums that users can use to
  verify that the downloads they receive have come from the Adoptium CI
  system. The verification process is defined in
  [this blog post](https://blog.adoptium.net/2022/07/gpg-signed-releases)
  and most of the process in there is applicable to the SBOMs for the latest
  releases.

- **Provenance - service generated**. The GPG signatures for the binary
  are generated as an integral part of the build process for the binaries
  and will be done in the same way for the signing of the SBOMs - the current

>   SBOM signing was done retrospectively in our Jenkins instance. The jenkins

Reviewer: ```suggestion
  SBOM signing was done retrospectively in our Jenkins instance. The Jenkins
```

---
title: SLSA level 2 compliance for Eclipse Temurin
date: "2022-11-11T12:00:00+00:00"
author: sxa
description: Eclipse Temurin by Adoptium is compliant with level 2 of the SLSA secure development framework.
tags:
  - Temurin
  - Security
---

# Introduction to SLSA

[SLSA](https://slsa.dev) is a framework with individual levels to indicate
increasing levels of security which software products can work towards. 
Today, we are pleased to announce that the Eclipse Temurin build and
distribution process is compliant with Level 2 of the SLSA specification.

What does this mean in practice? The [SLSA v0.1 levels are defined at this
link](http://slsa.dev/spec/v0.1/levels) and this means we are compliant with
levels 1 and 2 of the specification.

## SLSA Level 1

> Level 1 means that the build process is full scripted and automated and we

Reviewer 4: The first sentence is a run-on sentence. Recommend splitting it into at least two sentences.

are now producing Software Bill Of Material (SBOM) documents describing the
items which went into producing each of our builds which are published
alongside the downloads.  We have been compliant with level 1 for some time:
The level 1 requirements are:

- **Build - scripted build**: Our build steps are defined completely by the
  Jenkins pipelines from the [ci-jenkins-pipelines](https://github.com/adoptium/ci-jenkins-pipelines) repository and the
  underlying scripts run on the build machines from the
  [temurin-build](https://github.com/adoptium/temurin-build)
  repository - starting at
  [make-adopt-build-farm.sh](https://github.com/adoptium/temurin-build/blob/master/build-farm/make-adopt-build-farm.sh)

- **Provenance - available**: We create an SBOM in OWASP CycloneDX format
  along with our releases which contains all of the information about how

>   the builds are produced, which should allow someone to rebuild if required

Reviewer 4: Has this claim of rebuilding via SBOM details been tested thoroughly? Is it strictly necessary for meeting Level 1?

Author: No, which is why the wording is "should". I can remove it if you prefer.

  for any reason.  This includes the full set of parameters which we use to
  the makejdk_any_platform.sh script (invoked from
  make-adopt-build-farm.sh), the source repository tags (`scmRef`) which was
  used by the build process to produce the builds, the output from the
  openjdk `configure` invocation and various other pieces of information. 
  We are continually evolving the specific details which we include in the
  SBOM.  If you want to join in the discussion on the content, you can find
  the conversation in
  [temurin-build#3013](https://github.com/adoptium/temurin-build/issues/3013)
  or talk to us in the #secure-dev slack channel in the Adoptium workspace.

## SLSA Level 2

> Level 2 adds in additional requirements to provide some tamper resistance of

Reviewer 4: The first sentence here is also quite a lengthy read. Can it be split?

the build process, mostly in the areas of having all of our code version
controlled which we do by storing anything related to producing our buids in
git in our case and also by using a build service to produce and distribute
those builds - in our case this is our Jenkins CI server. In our release
cycles since July 2022 we have also provided PGP signature verification
information along with the SHA checksums that we previously had, which was
the final part of achieving SLSA level 2 compliance. The requirements here
are:

- **Source - Version controlled**. All of our source code is stored in GitHub in
  the jdkXX repositories such as
  [jdk17u](https://github.com/adoptium/jdk17u/), which we mirror from the
  [openjdk project](https://github.com/openjdk). These
  are version controlled and have tags for each release which we build from.
  Contributors agree to the rules of the
  [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

- **Build - Build service**. All of the build steps are run using our Jenkins

>   build service - currently at https://ci.adoptopenjdk.net which is used to

Reviewer 4: From:
```
service - currently at https://ci.adoptopenjdk.net which 
```

To:
```
service - currently at https://ci.adoptopenjdk.net - which 
```
or more concisely,
```
service (https://ci.adoptopenjdk.net) which 
```

  perform the builds, generate the SBOMs, and build the installers where
  applicable. The output from the builds are then posted into GitHub release
  repositories named as temurinXX-binaries (e.g.
  [temurin17-binaries](https://github.com/adoptium/temurin17-binaries) and
  also exposed via our API and web site.

- **Provenance - Authenticated**. We provide GPG checksums that users can use to
  verify that the downloads they receive have come from the Adoptium CI
  system. The verification process is defined in
  [this blog post](https://blog.adoptium.net/2022/07/gpg-signed-releases)

>   and most of the process in there is applicable to the SBOMs for the latest

Reviewer 4: ```
most of the process in there is applicable
```

What isn't applicable and why? That's the question that pops to mind after reading this statement.

Author: That statement will need to be changed as I'm looking at changing how they are distributed at the moment but to answer the specific question, the API calls to retrieve the data won't be as they are in those examples so they can't just be cut & pasted for this new scenario.

Author: Text of that section now adjusted (shortened!)

  releases.

- **Provenance - service generated**. The GPG signatures for the binary
  are generated as an integral part of the build process for the binaries
  and will be done in the same way for the signing of the SBOMs - the current
  SBOM signing was done retrospectively in our Jenkins instance. The Jenkins

>   instance hosts the private keys from the Eclipse Foundation which are used

Reviewer 4: This is already public but I found it odd to be informed of where the private keys are stored. Will the target audience care?

Author: This blog is about demonstrating compliance. The SLSA requirement is `The data in the provenance MUST be obtained from the build service` and it seemed appropriate to make that explicit.

>   to performing the signature generation.

Reviewer 4: `s/to performing/to perform`

# Next steps for the project

> Are we finished? Absolutely not! And we are continuing to work towards

Reviewer 4: `s/And we are continuing/We are continuing` reads better with the `and` that follows later on in the sentence.
